flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateUploader.java

```diff
@@ -107,13 +122,43 @@ private List<CompletableFuture<HandleAndLocalPath>> createUploadFutures(
-                                                                e,
-                                                                checkpointStreamFactory,
-                                                                stateScope,
-                                                                closeableRegistry,
-                                                                tmpResourcesRegistry)),
+                                                () -> {
+                                                    RetryStrategy retryStrategy =
+                                                            new FixedRetryStrategy(
+                                                                    DEFAULT_RETRY_TIMES,
+                                                                    DEFAULT_RETRY_DELAY);
+                                                    while (true) {
+                                                        try {
+                                                            return uploadLocalFileToCheckpointFs(
+                                                                    e,
+                                                                    checkpointStreamFactory,
+                                                                    stateScope,
+                                                                    closeableRegistry,
+                                                                    tmpResourcesRegistry);
+                                                        } catch (Throwable t) {
+                                                            if (retryStrategy
+                                                                            .getNumRemainingRetries()
+                                                                    == 0) {
+                                                                throw t;
+                                                            }
+
+                                                            LOG.warn(
+                                                                    "Fail to upload file to HDFS, will retry {} times, left retry times {}.",
+                                                                    DEFAULT_RETRY_TIMES,
+                                                                    retryStrategy
+                                                                            .getNumRemainingRetries(),
+                                                                    t);
+                                                        }
+
+                                                        TimeUnit.MILLISECONDS.sleep(
+                                                                retryStrategy
+                                                                        .getRetryDelay()
+                                                                        .toMillis());
```

Comment:
It would be better to not synchronously wait on the timeout, and free up the thread pool to do other things.